cpp/src/arrow/util/macros.h

```diff
@@ -97,6 +97,17 @@
 
 #endif  // ifndef NULLPTR
 
+// ----------------------------------------------------------------------
+// microsoft source-code annotation language
+
+#ifndef _In_z_
+#define _In_z_
+#endif
+
+#ifndef _Out_writes_opt_
+#define _Out_writes_opt_(size)
```

Comment:
Unless these macros can be introduced in a cross platform way I don't think we should add them to arrow.

Reply:
[These annotations are to support code analyzers for error checking](https://docs.microsoft.com/en-us/cpp/code-quality/annotating-function-parameters-and-return-values?view=msvc-170#annotations-and-descriptions) but AFAIK they are not necessary for building/running Arrow nor are standardized across platforms.

Reply:
these annotations are specific for msvc environment (required by prefast) and removed everywhere (where sal.h is not included) by these macros.

Reply:
I think modifying the codebase to support a static analyzer is a fairly disruptive change and the benefit should be carefully weighed against the maintenance overhead of these annotations. Could you please note on the JIRA why PREfast is a compelling analyzer to support?

I'm also curious about the reasoning for the specific locations where the annotations were added and whether PREfast could be profitably added to our CI suite